feat(user): Replaced the hardcoded return of a user’s solutions with the actual method (#406)

IsmaPeiro opened the pull request:
### Description:
User Story: [https://tree.taiga.io/project/luisvi-ita-challenges/us/394?milestone=456189](url)
This pull request removes the temporary hardcoded response for the getAllSolutionsByUser endpoint and replaces it with a fully functional implementation that retrieves data from MongoDB. Key updates include:

- **Repository Layer**: Introduced a method to query all user solutions by their UUID, enabling reactive access to stored documents.

- **Service Layer**: Updated the service interface to reflect the new repository-driven contract. In the implementation, the service now transforms retrieved documents into response DTOs, and throws a NotFoundException when no solutions are found.

- **Controller layer** : Enhanced the handler to emit informative logs for each retrieved solution and when no solutions are found. 
  
- **Unit Tests**: Revised existing tests to cover both scenarios—when no solutions exist (asserting the appropriate exception is thrown) and when multiple solutions are returned (verifying correct mapping to the response DTO).

IsmaPeiro committed:
added findAllByUserId on IUserSolutionRepository

IsmaPeiro committed:
Replaced temporary hardcoded UserSolutionServiceImpl.getAllSolutionsByUser implementation with real service method

IsmaPeiro committed:
Replaced test for temporary hardcoded method with real service method tests

IsmaPeiro committed:
added logs to getAllSolutionsByUser handler on UserController

IsmaPeiro committed:
removed unnecessary mock of UserServiceImpl on UserSolutionServiceImplTest

IsmaPeiro committed:
Removed the logs from the getAllSolutionsByUser handler in UserController; they will be included in another PR

IsmaPeiro committed:
added again logs to getAllSolutionsByUser handler on UserController

IsmaPeiro committed:
Update UserController.java

IsmaPeiro committed:
Update UserController.java

IsmaPeiro committed:
Update UserController.java

IsmaPeiro committed:
Update UserController.java

IsmaPeiro committed:
Update UserController.java

tonijimenez72 commented:
User has solutions: 200 - (Array of solutions)
![imatge](https://github.com/user-attachments/assets/9d8a0ccc-f408-468c-ab2d-830395f2a284)

User has no solutions: 404 - Solutions not found.
![imatge](https://github.com/user-attachments/assets/e0cd9285-161f-4172-b48b-d324dc000c73)

Invalid user ID: 404 - Invalid UUID string: (string)
![imatge](https://github.com/user-attachments/assets/b0bcfdef-9889-4b9a-b13e-1d2a426405ed)

![imatge](https://github.com/user-attachments/assets/b54047cf-c24c-484d-ba06-fd7f082eb6be)


- The endpoint specifies a 400 API response for "Invalid UUID", but it returns a 404.

- The test getAllSolutions_returns400IfInvalidUUID checks for a 400 response, but it should check for a 404 to align with the expected response.

- Should null or empty User UUID be controlled?

The code is well-structured, clean, and easy to understand. Good job!

IsmaPeiro committed:
refactor(user): extract UUID validation for getAllSolutionsByUser into reactive helper and add unit tests

IsmaPeiro commented:
> User has solutions: 200 - (Array of solutions) ![imatge](https://private-user-images.githubusercontent.com/156689808/447892265-9d8a0ccc-f408-468c-ab2d-830395f2a284.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDgzNjI4ODMsIm5iZiI6MTc0ODM2MjU4MywicGF0aCI6Ii8xNTY2ODk4MDgvNDQ3ODkyMjY1LTlkOGEwY2NjLWY0MDgtNDY4Yy1hYjJkLTgzMDM5NWYyYTI4NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyN1QxNjE2MjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNGE4ZDNkOTA4MzgwOTA0MWRjOGQ1NGY1ZjJiNTk1NGI3YzcxZWE3ODI1NWM4OTg0NDYwZGQ0MmRmOGYyODAzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.6IIO0tHA5NiXE5RhJuL-xwWYjPDKWXSzeUTrC8h4GD8)
> 
> User has no solutions: 404 - Solutions not found. ![imatge](https://private-user-images.githubusercontent.com/156689808/447891138-e0cd9285-161f-4172-b48b-d324dc000c73.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDgzNjI4ODMsIm5iZiI6MTc0ODM2MjU4MywicGF0aCI6Ii8xNTY2ODk4MDgvNDQ3ODkxMTM4LWUwY2Q5Mjg1LTE2MWYtNDE3Mi1iNDhiLWQzMjRkYzAwMGM3My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyN1QxNjE2MjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xM2RkYTY3MGIwMjhiNmZhNGRhNzY4MGVmOWZjODRlMGQ3NjJjYWNiYjU0NjQ2ODJkNDQ2NDEzNzhjYTZjZjhmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.d8S3IOsIRcXGZglc01PKGAD2yvPiZET78qvL0ZyMFfg)
> 
> Invalid user ID: 404 - Invalid UUID string: (string) ![imatge](https://private-user-images.githubusercontent.com/156689808/447895292-b0bcfdef-9889-4b9a-b13e-1d2a426405ed.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDgzNjI4ODMsIm5iZiI6MTc0ODM2MjU4MywicGF0aCI6Ii8xNTY2ODk4MDgvNDQ3ODk1MjkyLWIwYmNmZGVmLTk4ODktNGI5YS1iMTNlLTFkMmE0MjY0MDVlZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyN1QxNjE2MjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZWJkZTAwYTEwODBkZjk1YTI5ODAyOGQyZWZjMmEwZDkyNTJiMTYyNGY0N2VlMDA5NGEzNmM0MDVlMzQ4ODIzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.-unavYbQx_8ObKorhq1csc0HB7hTmlTuViCI0D2S6Fo)
> 
> ![imatge](https://private-user-images.githubusercontent.com/156689808/447898349-b54047cf-c24c-484d-ba06-fd7f082eb6be.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDgzNjI4ODMsIm5iZiI6MTc0ODM2MjU4MywicGF0aCI6Ii8xNTY2ODk4MDgvNDQ3ODk4MzQ5LWI1NDA0N2NmLWMyNGMtNDg0ZC1iYTA2LWZkN2YwODJlYjZiZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyN1QxNjE2MjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kODcwZTg1MWU0MmM4ZTQ1NjdjMWUzMmFiZDU3NWUzYTJkNzVlZGUzODZmZGQ3MjdiN2Q1YmFkYjYzM2UwNjdlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.d7-EtR0uzmtBSqid3TzqLnQYV7Ap0-tAuLd4_NWhqA8)
> 
> * The endpoint specifies a 400 API response for "Invalid UUID", but it returns a 404.
> * The test getAllSolutions_returns400IfInvalidUUID checks for a 400 response, but it should check for a 404 to align with the expected response.
> * Should null or empty User UUID be controlled?
> 
> The code is well-structured, clean, and easy to understand. Good job!

Hi @tonijimenez72, thanks for the feedback! I’ve made the following changes in response:

I extracted all of the userId checks into a private reactive helper that immediately returns a BadRequestException (400) if the ID is null, empty, or not a valid UUID. This means invalid inputs no longer reach the repository.

I kept the existing behavior of throwing a NotFoundException (404) only when the repository actually returns an empty result, so “no solutions” remains a 404.

I expanded the unit tests to cover every edge case in the public API:

Null, empty, and malformed userId inputs now each trigger a 400 without calling the repository.

I’ve changed the API so that an invalid UUID now returns a 400 Bad Request, and therefore the test getAllSolutions_returns400IfInvalidUUID is now asserting the correct, intended behavior.

![image](https://github.com/user-attachments/assets/e6e3b1d9-abe2-495c-814f-2e30320cc1fa)

With these updates, the API now consistently returns 400 for bad IDs and 404 only for genuinely missing data, and our test suite fully documents that behavior. Let me know if you’d like any further tweaks!

tonijimenez72 commented:
Null or empty returns 400.
![imatge](https://github.com/user-attachments/assets/868f755f-d035-415c-ab26-00e4718c8a3d)
Bad request returns 400.
![imatge](https://github.com/user-attachments/assets/f813e337-46d9-4e51-9b28-4003da0a0cfd)

tonijimenez72 reviewed (approved):
Reviewed and approved.

IsmaPeiro committed:
Added CHANGELOG.md entry

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/controller/UserController.java:
```diff
@@ -423,7 +423,14 @@ public Mono<ResponseEntity<Flux<UserSolutionResponseDto>>> getAllSolutionsByUser
             @PathVariable String userId
```
👀 I see that the GET /users/{userId}/solutions endpoint exposes a specific user's solutions.
Have you considered whether this endpoint should be secured?

Although it's a GET, it returns potentially sensitive information (the user's solutions), so it may make sense to require authentication and validate that the authenticated user has permission to access that data (e.g., they're the user themselves or have a mentor/admin role).

🔐 In other endpoints like POST /logout, you're already applying validations to the JWT token, so you could follow similar logic here if you see fit.

What was your approach?

IsmaPeiro replied:
I followed the existing pattern in UserController by leaving its endpoints unsecured, but I agree it would be wise to secure them all to avoid exposing sensitive user data. Since this wasn’t part of the original user-story requirements and given our time constraints to meet the sprint deadline, I think it’s more practical to log this as technical debt and tackle it in a future iteration. If you consider it critical, I can try to reprioritize and secure it right away.

Action: Log it in the backlog as technical debt to secure UserController endpoints.

ivilarop replied:
No, don't worry, it's fine to audit it as technical debt, and we'll address it later. But keep this in mind for future developments, since we should have initially detected it in the story and marked it as optional. We can close the conversation. Thank you.

IsmaPeiro replied:
I’ll keep it in mind, thank you very much.

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/controller/UserController.java:
```diff
@@ -423,7 +423,14 @@ public Mono<ResponseEntity<Flux<UserSolutionResponseDto>>> getAllSolutionsByUser
-                .body(userSolutionService.getAllSolutionsByUser(userId))
+                .body(userSolutionService.getAllSolutionsByUser(userId)
+                        .doOnNext(dto ->
+                                log.info("→ Solution retrieved for user {}: challengeId={}", userId, dto.getChallengeId()))
```
👀 In the ChallengeController, I see you're using .doOnNext(...) to execute a log.info(...) with the userId and challengeId.

🔁 This log can be useful for traceability, but conceptually it would be more appropriate to move it to the service, which is where you're already building the DTO and applying the business logic.

📉 Additionally, since it's in the controller (and if multiple solutions are issued), this log.info will be executed for each element of the Flux, which can generate a lot of noise in the logs if the user has many solutions or if the request is repeated frequently (for example, during a UI refresh).

✅ Moving this log to the service's map() allows you to centralize traceability and avoid overloading the controller with responsibilities that don't correspond to it.

IsmaPeiro replied:
I’ve removed the .doOnNext(…) from the controller and moved the log.info("→ Solution retrieved…") into the service’s .map(…) when building each DTO. This keeps the controller focused on HTTP handling and centralizes traceability with the business logic.

ivilarop replied:
ok thks!

IsmaPeiro replied:
Thanks to you.

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/controller/UserController.java:
```diff
@@ -423,7 +423,14 @@ public Mono<ResponseEntity<Flux<UserSolutionResponseDto>>> getAllSolutionsByUser
+                .body(userSolutionService.getAllSolutionsByUser(userId)
+                        .doOnNext(dto ->
+                                log.info("→ Solution retrieved for user {}: challengeId={}", userId, dto.getChallengeId()))
+                        .switchIfEmpty(Flux.defer(() -> {
```
In the controller, you're also using .switchIfEmpty(...) to log a log.warn() if no solutions are found.

However, you're already handling the empty case within the userSolutionService itself.

👉 Therefore, this switchIfEmpty in the controller is not only redundant, but it can cause confusion by having two different logics for the same case (one throws an exception, the other returns empty).

💡 I would suggest keeping that responsibility solely in the service and removing it from the controller to avoid duplication and ambiguity.

IsmaPeiro replied:
I’ve eliminated the switchIfEmpty(...) (and its warning) from the controller. All empty-case behavior is now solely in the service layer.

ivilarop replied:
ok thks!

IsmaPeiro replied:
Thanks to you.

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/service/UserSolutionServiceImpl.java:
```diff
@@ -81,25 +83,30 @@ private Mono<UserSolutionDocument> saveValidSolution(UUID userUuid, UUID challen
+                .flatMapMany(uuid ->
+                userSolutionRepository
+                        .findAllByUserId(UUID.fromString(userId))
+                        .switchIfEmpty(Mono.error(new NotFoundException("Solutions not found."))))
```
👀 Currently, in userSolutionService.getAllSolutionsByUser(...), you are throwing a NotFoundException when no solutions are found.

I'd like you to consider whether this behavior is the most appropriate:

Does it make sense to consider the fact that a user hasn't yet submitted solutions as an "error"?

Or would it be more correct (and consistent with REST APIs) to return a 200 OK with an empty list?

You might also consider whether this decision (returning 404 vs. 200) should be made from the controller or from the service.
I encourage you to do some research and justify your choice.

IsmaPeiro replied:
I was just debating this with our frontend colleague this morning.
By REST convention, a user with zero solutions should get 200 OK + an empty list, not a 404. I’ve updated the service to remove the switchIfEmpty(Mono.error(...)), so findAllByUserId now returns a Flux.empty(). The controller naturally serializes that to [].

Test update: The unit test now uses expectNextCount(0) and verifyComplete() instead of expecting a NotFoundException.

ivilarop replied:
ok thks!

IsmaPeiro replied:
Thanks to you.

IsmaPeiro committed:
getAllSolutionsByUser now returns empty array when no solutions found

IsmaPeiro committed:
updated getAllSolutionsByUser_noSolutionsFound_test

IsmaPeiro committed:
removed logs from getAllSolutionsByUser on UserController
